File: license-report/src/main/java/org/abhijitsarkar/licensereport/GitLabProperties.java
package org.abhijitsarkar.licensereport;

import lombok.Data;
import org.springframework.boot.context.properties.ConfigurationProperties;

import java.util.ArrayList;
import java.util.List;

/**
 * @author Abhijit Sarkar
 */
@ConfigurationProperties("gitlab")
@Data
public class GitLabProperties {
    private String baseUrl;
    private List<GroupProperties> groups = new ArrayList<>();
    private ConnectionProperties connection = new ConnectionProperties();

    @Data
    public static class GroupProperties {
        private String name;
        private String privateToken;
        private List<String> excludesProjects = new ArrayList<>();
        private List<String> includesProjects = new ArrayList<>();
    }

    @Data
    public static class ConnectionProperties {
        private int connectTimeoutMillis = 500;
        private int readTimeoutMillis = 5000;
    }
}

File: license-report-kotlin/src/main/kotlin/org/abhijitsarkar/client/GitLabProperties.kt
package org.abhijitsarkar.client

import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.boot.context.properties.NestedConfigurationProperty

/**
 * @author Abhijit Sarkar
 */
@ConfigurationProperties("licensereport.gitlab")
class GitLabProperties {
    lateinit var baseUrl: String

    @NestedConfigurationProperty
    val groups = mutableMapOf<String, GroupProperties>()

    @NestedConfigurationProperty
    val connection = ConnectionProperties()

    class GroupProperties {
        lateinit var privateToken: String
        val excludedProjects = mutableMapOf<String, ProjectProperties>()
        val includedProjects = mutableMapOf<String, ProjectProperties>()
    }

    class ProjectProperties {
        var branch = "master"
    }

    class ConnectionProperties {
        var connectTimeoutMillis = 500
        var readTimeoutMillis = 5000
    }
}
